### PR TITLE
feat(docker): shift app to docker

### PR DESCRIPTION
**Pull request by seaerchin**
## Problem
We want to shift our infra over to ecs instead of eb as ecs is faster to deploy + easier to work with. This PR dockerises our app as a first step. 

## Solution
1. add docker file
2. add docker compose

## Notes
- hot reload is enabled still by using volume mapping; we map the vol on disk to the docker volume so that on chnage of the files on disk, docker notices
- as our containers are part of the same network, the postgres url has to use the docker volume instead. (see comment for format in code)

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,7 @@
+FROM node:18-alpine AS base
+WORKDIR /opt/isomercms-backend
+COPY . .
+RUN npm ci
```
Comment (harishv7) [inline Dockerfile:4]: Getting this err for npm ci

``` > [4/5] RUN npm ci:
#0 14.67 npm WARN deprecated w3c-hr-time@1.0.2: Use your platform's native performance.now() and performance.timeOrigin.
#0 14.70 npm WARN deprecated @npmcli/move-file@1.1.2: This functionality has been moved to @npmcli/fs
#0 15.75 npm WARN deprecated querystring@0.2.0: The querystring API is considered Legacy. new code should use the URLSearchParams API instead.
#0 15.88 npm WARN deprecated uuid@3.4.0: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.
#0 15.92 npm WARN deprecated formidable@1.2.6: Please upgrade to latest, formidable@v2 or formidable@v3! Check these notes: https://bit.ly/2ZEqIau
#0 17.20 npm WARN deprecated superagent@3.7.0: Please upgrade to v7.0.2+ of superagent.  We have fixed numerous issues with streams, form-data, attach(), filesystem errors not bubbling up (ENOENT on attach()), and all tests are now passing.  See the releases tab for more information at <https://github.com/visionmedia/superagent/releases>.
#0 23.64 npm notice
#0 23.64 npm notice New patch version of npm available! 10.2.3 -> 10.2.5
#0 23.64 npm notice Changelog: <https://github.com/npm/cli/releases/tag/v10.2.5>
#0 23.64 npm notice Run `npm install -g npm@10.2.5` to update!
#0 23.64 npm notice
#0 23.64 npm ERR! code ECONNRESET
#0 23.64 npm ERR! network aborted
#0 23.64 npm ERR! network This is a problem related to network connectivity.
#0 23.64 npm ERR! network In most cases you are behind a proxy or have bad network settings.
#0 23.66 npm ERR! network
#0 23.66 npm ERR! network If you are behind a proxy, please make sure that the
#0 23.66 npm ERR! network 'proxy' config is set properly.  See: 'npm help config'
#0 23.66
#0 23.66 npm ERR! A complete log of this run can be found in: /root/.npm/_logs/2023-12-15T05_28_14_844Z-debug-0.log
------
failed to solve: executor failed running [/bin/sh -c npm ci]: exit code: 1
```

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,7 @@
...
+EXPOSE "8081"
+COPY . /opt/isomercms-backend
```
Comment (harishv7) [inline Dockerfile:6]: what's this copy for? there was already a copy on line 3

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,7 @@
...
+CMD ["npm", "run", "dev:server"]
```
Comment (harishv7) [inline Dockerfile:7]: how are env vars being injected in this?

File: package.json [hunk 4]
```diff
@@ -5,9 +5,8 @@
   "scripts": {
     "build": "tsc -p tsconfig.build.json",
     "start": "node --unhandled-rejections=warn -r ts-node/register/transpile-only -r tsconfig-paths/register -r dotenv/config build/server.js dotenv_config_path=/efs/isomer/.isomer.env",
-    "dev:services": "docker compose up -d",
     "dev:server": "source .env && ts-node-dev --unhandled-rejections=warn --respawn src/server.js",
-    "dev": "npm run dev:services && npm run dev:server",
+    "dev": "docker compose up",
```
Comment (harishv7) [inline package.json:9]: how about build and start?

Comment (seaerchin) [reply package.json:9]: that's for serving our _production_ site - no need to change

File: Dockerfile (hunk 3 above)
Comment (seaerchin) [reply Dockerfile:7]: docker files don't actually do this - this is done inside `docker-compose.yml` through `env-file` and `environment`

File: Dockerfile (hunk 2 above)
Comment (seaerchin) [reply Dockerfile:6]: removing the one on line 3, is extra

File: Dockerfile (hunk 1 above)
Comment (seaerchin) [reply Dockerfile:4]: it seems l ike a network issue and not a docker issue?

Comment (harishv7) [reply Dockerfile:4]: if it is working fine for other engs, then all good

File: package.json (hunk 4 above)
Comment (harishv7) [reply package.json:9]: ah this is primarily for dev only? ok then

Comment (seaerchin) [reply package.json:9]: yea, on prod we need to write a deploy file to ecr, that will be consuemd to ecs

File: Dockerfile (hunk 1 above)
Comment (seaerchin) [reply Dockerfile:4]: @dcshzj @QiluXie can try?

Comment (dcshzj) [reply Dockerfile:4]: I'm getting different errors:
1. `Unable to query tokens from database: SequelizeConnectionRefusedError: connect ECONNREFUSED 127.0.0.1:5432`
2. `Invalid GitHub response format: [object Object]`

Not able to log in at all :(

Comment (dcshzj) [reply Dockerfile:4]: Fixed by removing the `DB_URI` env var, still pending EFS fix

Comment (seaerchin) [reply Dockerfile:6]: resolved, mounted efs vol

**Review by harishv7 — APPROVED** [review]

**Pull request merged**